From 8d165d5fcb4fe79da96683bf13672e835f95705b Mon Sep 17 00:00:00 2001
From: Joyce McIntosh <joyce.mcintosh@nexenta.com>
Date: Wed, 16 May 2018 13:03:07 -0600
Subject: [PATCH] NEX-17125 NFS: nbmand lock entered but not exited on error
 path Reviewed by: Yuri Pankov <yuri.pankov@nexenta.com> Reviewed by: Rick
 McNeal <rick.mcneal@nexenta.com> Reviewed by: Sanjay Nadkarni
 <sanjay.nadkarni@nexenta.com> Reviewed by: Evan Layton
 <evan.layton@nexenta.com> Reviewed by: Gordon Ross <gordon.ross@nexenta.com>

---
 usr/src/uts/common/fs/nfs/nfs_srv.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/usr/src/uts/common/fs/nfs/nfs_srv.c b/usr/src/uts/common/fs/nfs/nfs_srv.c
index fc802b8049..eb1c7e775b 100644
--- a/usr/src/uts/common/fs/nfs/nfs_srv.c
+++ b/usr/src/uts/common/fs/nfs/nfs_srv.c
@@ -799,6 +799,8 @@ rfs_read(struct nfsreadargs *ra, struct nfsrdresult *rr,
 
 	/* check if a monitor detected a delegation conflict */
 	if (error == EAGAIN && (ct.cc_flags & CC_WOULDBLOCK)) {
+		if (in_crit)
+			nbl_end_crit(vp);
 		VN_RELE(vp);
 		/* mark as wouldblock so response is dropped */
 		curthread->t_flag |= T_WOULDBLOCK;
@@ -1124,10 +1126,7 @@ rfs_write_sync(struct nfswriteargs *wa, struct nfsattrstat *ns,
 
 	/* check if a monitor detected a delegation conflict */
 	if (error == EAGAIN && (ct.cc_flags & CC_WOULDBLOCK)) {
-		VN_RELE(vp);
-		/* mark as wouldblock so response is dropped */
-		curthread->t_flag |= T_WOULDBLOCK;
-		return;
+		goto out;
 	}
 
 	if (wa->wa_data || wa->wa_rlist) {
-- 
2.17.2 (Apple Git-113)

